### PR TITLE
feat(mobile): ios oauth & magic-link login

### PR DESCRIPTION
**Pull request by CatsJuice**

### CODE REVIEW

**Comment by CatsJuice** [comment]
* **#8581** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8581?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @CatsJuice and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/core/src/modules/cloud/services/auth.ts [hunk 1]
```diff
@@ -116,13 +116,18 @@ export class AuthService extends Service {
   ) {
     track.$.$.auth.signIn({ method: 'magic-link' });
     try {
+      const scheme = BUILD_CONFIG.isElectron
+        ? appInfo?.schema
+        : BUILD_CONFIG.isIOS || BUILD_CONFIG.isAndroid
+          ? 'affine'
+          : 'web';
```
Comment (pengx17) [inline packages/frontend/core/src/modules/cloud/services/auth.ts]: Are we planning to submit different app versions for canary/beta/stable?

Comment (CatsJuice) [reply packages/frontend/core/src/modules/cloud/services/auth.ts]: Scheme in electron is dynamic for each channel, but it's fixed for iOS. Not sure about android yet.

File: packages/frontend/core/package.json [hunk 2]
```diff
@@ -18,6 +18,8 @@
     "@affine/track": "workspace:*",
     "@blocksuite/affine": "0.17.19",
     "@blocksuite/icons": "2.1.69",
+    "@capacitor/app": "^6.0.1",
```
Comment (EYHN) [inline packages/frontend/core/package.json:21]: I think we should avoid referencing the capacitor package in core

File: packages/frontend/core/src/components/affine/auth/send-email.tsx [hunk 3]
```diff
@@ -122,7 +122,11 @@ const useSendEmail = (emailType: AuthPanelProps<'sendEmail'>['emailType']) => {
         return trigger({
           email,
           callbackUrl: `/auth/${callbackUrl}?isClient=${
```
Comment (EYHN) [inline packages/frontend/core/src/components/affine/auth/send-email.tsx]: AppSchema is not needed here?

File: packages/frontend/core/src/utils/popup.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import { DebugLogger } from '@affine/debug';
 import { apis } from '@affine/electron-api';
+import { Browser } from '@capacitor/browser';
 
 const logger = new DebugLogger('popup');
```
Comment (EYHN) [inline packages/frontend/core/src/utils/popup.ts]: <i class='graphite__hidden'>[Re: line +15]<br/><br/></i>

We need to abstract this into a popupWindowProvider<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8581?utm_source=unchanged-line-comment">Graphite</a>.</span>

File: packages/frontend/core/src/components/affine/auth/send-email.tsx (hunk 3 above)
Comment (CatsJuice) [reply packages/frontend/core/src/components/affine/auth/send-email.tsx]: Seems `client` is not used in `/auth/:authType`

File: packages/frontend/core/package.json (hunk 2 above)
Comment (CatsJuice) [reply packages/frontend/core/package.json:21]: but `popupWindow` is implemented in core. Should we wrap a function in `@affine/ios` and import from it?

File: packages/frontend/core/src/modules/url/services/url.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import type { ClientSchemaProvider } from '../providers/client-schema';
+import type { PopupWindowProvider } from '../providers/popup-window';
+
+export class UrlService extends Service {
+  // those providers are optional, because they are not always available in some environments
+
+  private readonly popupWindowProvider: PopupWindowProvider | undefined;
+  private readonly clientSchemaProvider: ClientSchemaProvider | undefined;
+  constructor(
+    popupWindowProvider: PopupWindowProvider[],
```
Comment (pengx17) [inline packages/frontend/core/src/modules/url/services/url.ts]: why define popupWindowProvider as an array but only the first value is used?

**Review by pengx17 — APPROVED** [review]

**Pull request merged**